### PR TITLE
partition videos across SLURM array tasks

### PR DESCRIPTION
**Pull request by Hilda-Azimi**
## Notes for review

- `num_partitions` and the `--array` upper bound must be kept in sync by
  hand. Nothing at runtime can detect a mismatch.
- No resume logic — a requeued node re-encodes its whole partition.
  Partitions never collide, so it's safe, just wasteful.
- `parallel_compression` is removed. The docker example in the README
  still passes it.
- The commented-out `_run_parallel` / `parallel_compression` blocks are
  left in for context; happy to delete them if you'd prefer.

### CODE REVIEW

File: src/aind_behavior_video_transformation/etl.py [hunk 1]
```diff
@@ -3,7 +3,11 @@
 import logging
 import shlex
 import sys
-from concurrent.futures import ProcessPoolExecutor, as_completed
+
+# CHANGED: ProcessPoolExecutor / as_completed are no longer used. Parallelism
```
Comment (jtyoung84) [inline src/aind_behavior_video_transformation/etl.py]: It might be cleaner to put these changes in the release notes instead of as comments in code.

File: src/aind_behavior_video_transformation/etl.py [hunk 2]
```diff
@@ -201,6 +232,34 @@ def run_job(self) -> JobResponse:
...
+        # this same list independently and then takes its own stride, so all
+        # nodes must agree on the ordering. transform_directory does not
+        # guarantee one.
+        convert_video_args.sort(key=lambda x: str(x[0]))
+
+        total_videos = len(convert_video_args)
+        partition_number = self.job_settings.partition_number
+        num_partitions = self.job_settings.num_partitions
+
+        if partition_number > num_partitions:
```
Comment (jtyoung84) [inline src/aind_behavior_video_transformation/etl.py:336]: nit: we can consider validating this in the `job_settings` class. Pydantic models have a validation API. It will help raise the error earlier.

File: .gitignore [hunk 3]
```diff
@@ -146,3 +146,4 @@ dmypy.json
 *.flv
 *.mkv
 *.webm
+uv.lock
```
Comment (jtyoung84) [inline .gitignore]: Instead of adding the `uv.lock` to the `.gitignore` file, it's recommended to create a `.gitattributes` file and add this line:
```
uv.lock   merge=binary linguist-generated
```

File: src/aind_behavior_video_transformation/etl.py [hunk 4]
```diff
@@ -201,6 +232,34 @@ def run_job(self) -> JobResponse:
...
+            raise ValueError(
+                f"partition_number ({partition_number}) exceeds "
+                f"num_partitions ({num_partitions})"
+            )
+
+        convert_video_args = convert_video_args[
```
Comment (jtyoung84) [inline src/aind_behavior_video_transformation/etl.py]: We might need to add custom partition logic to ensure the video files are distributed evenly.

File: src/aind_behavior_video_transformation/etl.py (hunk 1 above)
Comment (Hilda-Azimi) [reply src/aind_behavior_video_transformation/etl.py]: All comments are removed.

File: .gitignore (hunk 3 above)
Comment (Hilda-Azimi) [reply .gitignore]: Done, added .gitattributes and removed the .gitignore entry so uv.lock is tracked.

File: src/aind_behavior_video_transformation/etl.py (hunk 4 above)
Comment (Hilda-Azimi) [reply src/aind_behavior_video_transformation/etl.py]: Should I add it?

**Comment by Hilda-Azimi** [comment]
Thanks for the review.

- uv.lock: removed from .gitignore and now tracked. Added a .gitattributes at the repo root with uv.lock merge=binary linguist-generated.

- Inline change notes: removed all.

- Even distribution: I assumed you meant balancing work rather than file count. The old slice already balanced counts. If you meant something else, please let me know.

File: src/aind_behavior_video_transformation/etl.py (hunk 4 above)
Comment (Hilda-Azimi) [reply src/aind_behavior_video_transformation/etl.py]: Should I add it?

File: src/aind_behavior_video_transformation/etl.py [hunk 5]
```diff
@@ -163,6 +170,60 @@ def _run_compression(
             f"{len(errors)} ffmpeg job(s) failed:\n\n" + "\n\n".join(formatted)
         )
 
+    @staticmethod
+    def _partition_args(
```
Comment (jtyoung84) [inline src/aind_behavior_video_transformation/etl.py:229]: It might be helpful to add a few unit tests in the testing library to test this specific method with a few edge cases.

**Comment by Hilda-Azimi** [comment]
I ran the tests and got a failure. I checked whether it was related to my branch, but I get the same failure on main, so it was pre-existing. It's in build_overrides_dict. It was fine on CI because CI runs Linux. It only shows up on macOS. The reason is that macOS makes /var a symlink to /private/var, so a temp folder ends up with two valid names for the same place.

File: src/aind_behavior_video_transformation/etl.py [hunk 6]
```diff
@@ -163,6 +170,64 @@ def _run_compression(
...
+            else:
+                others.append(params)
+
+        partitions: list[list] = [[] for _ in range(num_partitions)]
+
+        loads = [0] * num_partitions
+        weighted = sorted(
+            ((params[0].stat().st_size, params) for params in videos),
+            key=lambda pair: (-pair[0], str(pair[1][0])),
+        )
```
Comment (galenlynch) [inline src/aind_behavior_video_transformation/etl.py:273]: I am not sure how important this ranking is, but number of frames would likely be a better proxy for runtime than file size. You can use `ffprobe` to get that. Also it looks like you're negating the filesize to reverse the ranking? If so, then using `reverse=True` would be more clear.

File: src/aind_behavior_video_transformation/etl.py [hunk 7]
```diff
@@ -78,6 +73,44 @@ class BehaviorVideoJobSettings(BasicJobSettings):
...
+            ".flv",
+            ".m4v",
+            ".webm",
+        },
+        description=(
+            "Lowercase suffixes treated as videos when balancing partitions. "
+            "Entries not matching are assumed to be symlinked at negligible "
+            "cost and are spread by count rather than by weight."
+        ),
+    )
```
Comment (galenlynch) [inline src/aind_behavior_video_transformation/etl.py:167]: This is duplicated in `filesystem.likely_video_file`. Please consolidate it.

File: src/aind_behavior_video_transformation/etl.py [hunk 8]
```diff
@@ -201,6 +266,40 @@ def run_job(self) -> JobResponse:
...
+            raise ValueError(
+                f"partition_number ({partition_number}) exceeds "
+                f"num_partitions ({num_partitions})"
+            )
+
+        convert_video_args = self._partition_args(
+            convert_video_args,
+            num_partitions,
+            video_extensions,
+        )[partition_number - 1]
```
Comment (galenlynch) [inline src/aind_behavior_video_transformation/etl.py]: Could you add a comment explaining what this does, and maybe not shadow the input `convert_video_args` with the output? `_partition_args` splits up the arguments list into partitions, right? You're selecting the jobs that will run on _this_ partition?

I'm actually a bit confused about the execution context here. Each partition will run this code? It might be helpful to explain that so readers with less familiarity with the data-transfer-services can understand the code.

File: src/aind_behavior_video_transformation/etl.py (hunk 6 above)
Comment (Hilda-Azimi) [reply src/aind_behavior_video_transformation/etl.py:273]: Good points, thanks. I applied your comment.

File: src/aind_behavior_video_transformation/etl.py (hunk 7 above)
Comment (Hilda-Azimi) [reply src/aind_behavior_video_transformation/etl.py:167]: I fixed it in both .py files.

File: src/aind_behavior_video_transformation/etl.py [hunk 9]
```diff
@@ -46,6 +46,25 @@ def _format_ffmpeg_error(video_path: Path, exc: CalledProcessError) -> str:
...
+    out = subprocess.run(
+        ["ffprobe", "-v", "error", "-select_streams", "v:0",
+         "-count_frames", "-show_entries", "stream=nb_read_frames",
+         "-of", "csv=p=0", str(path)],
+        check=False, capture_output=True, text=True,
+    )
+    text = out.stdout.strip()
+    if not text or text == "N/A":
+        return Path(path).stat().st_size
+    return int(text)
```
Comment (galenlynch) [inline src/aind_behavior_video_transformation/etl.py]: For most videos, you don't need to count frames explicitly, which is what `nb_read_frames` does. Doing so will force the process to decode the entire video to count the frames. Instead, you can use the `nb_frames` stream which returns what the stream header has, and does not decode frames. For videos that do not have a `nb_frames` count (on-pipeline videos should all have this), you can fall back to decoding frames to count them.

Total pixel count would also allow for different resolutions to be correctly load balanced.

If neither works, I think conflating the frame count with file size will lead to surprising and wrong results. If `ffprobe` cannot count the frames it will be because the video is corrupt. If the video is corrupt I think failing loudly is more appropriate than returning a file size number that is not directly comparable to the frame counts.

Suggested rewrite below:
```suggestion
def encode_cost(path: Path) -> int:
    """Return total video pixels as a proxy for encode cost.

    Uses container metadata when available, falling back to scanning the
    video stream only when the container does not provide a frame count.
    """
    path = Path(path)

    out = subprocess.run(
        [
            "ffprobe",
            "-v", "error",
            "-select_streams", "v:0",
            "-show_entries", "stream=nb_frames,width,height",
            "-of", "json",
            str(path),
        ],
        check=True,
        capture_output=True,
        text=True,
    )

    stream = json.loads(out.stdout)["streams"][0]
    width = int(stream["width"])
    height = int(stream["height"])

    frame_count = stream.get("nb_frames")
    if frame_count not in (None, "N/A"):
        return int(frame_count) * width * height

    # Some containers do not expose nb_frames, so count them explicitly.
    out = subprocess.run(
        [
            "ffprobe",
            "-v", "error",
            "-select_streams", "v:0",
            "-count_frames",
            "-show_entries", "stream=nb_read_frames",
            "-of", "default=noprint_wrappers=1:nokey=1",
            str(path),
        ],
        check=True,
        capture_output=True,
        text=True,
    )

    frame_count = out.stdout.strip()
    if not frame_count or frame_count == "N/A":
        raise ValueError(f"Could not determine frame count for {path}")

    return int(frame_count) * width * height
```

File: src/aind_behavior_video_transformation/etl.py (hunk 8 above)
Comment (Hilda-Azimi) [reply src/aind_behavior_video_transformation/etl.py]: Good call, done. Renamed the output to `this_partition` so it no longer shadows the input `convert_video_args`, and added a comment explaining the execution model. Each partition runs this same script independently.

File: src/aind_behavior_video_transformation/etl.py (hunk 9 above)
Comment (Hilda-Azimi) [reply src/aind_behavior_video_transformation/etl.py]: Thanks, these are all fair points. One thing I want to work out before adopting the rewrite: the partitioning tests currently use synthetic fixtures (zero-byte stubs and non-existent paths) to exercise the balancing logic in isolation, without depending on ffmpeg or real video files. Since `encode_cost` raises on anything that isn't a real decodable video, those tests will fail as written — the file-size fallback was what kept them green. So moving to this approach means either generating real (tiny) video fixtures, or decoupling the cost function from the balancing tests some other way. I want to think about which is the cleanest path before I push a change — I'll follow up. Leaning toward real fixtures so the tests actually validate the pixel/frame-count logic, but want to make sure that doesn't make the suite slow or flaky first.

Comment (galenlynch) [reply src/aind_behavior_video_transformation/etl.py]: There is already a video in the repo for testing: [tests/test_video_in_dir/clip.mp4](https://github.com/AllenNeuralDynamics/aind-behavior-video-transformation/blob/main/tests/test_video_in_dir/clip.mp4)

Comment (galenlynch) [reply src/aind_behavior_video_transformation/etl.py]: You could also mock the ffprobe `subprocess.run` calls, but IMO doing an integration test with ffprobe should be required. One alternative would be to explicitly separate out the probing code so it would be easier to mock in `encode_cost`. Even in that case, an integration test should be required.

```python
def probe_video(path: Path) -> VideoInfo:
    """Call ffprobe and return video metadata."""
    ...


def encode_cost(info: VideoInfo) -> int:
    """Return an encode-cost estimate from video metadata."""
    return info.frames * info.width * info.height
```

Comment (jtyoung84) [reply src/aind_behavior_video_transformation/etl.py]: I'm a little bit worried about scope creep. The original request was to just partition the video files. I'd rather just implement that, and, if we really need to rank video files by size or frame count, then do that in a future PR if the need arises. If this isn't too complicated to implement now, then let's go ahead and finish it, but this additional feature shouldn't block us from deploying things soon.

Comment (galenlynch) [reply src/aind_behavior_video_transformation/etl.py]: I would be fine with removing the file size load balancing as well. My original comment was that file size is not the best proxy for encoding time.

**Review by galenlynch — COMMENTED** [review]
Looks great to me, but I'll leave the final approval to @jtyoung84!

File: src/aind_behavior_video_transformation/etl.py [hunk 10]
```diff
@@ -78,6 +139,33 @@ class BehaviorVideoJobSettings(BasicJobSettings):
...
+    num_partitions: int = Field(
+        default=1,
+        ge=1,
+        description=(
+            "Total number of partitions in the array. Must be identical on "
+            "every node. Default 1 means this node processes all videos."
+        ),
+    )
+
+    video_extensions: set[str] = Field(
```
Comment (jtyoung84) [inline src/aind_behavior_video_transformation/etl.py:160]: We probably don't need this as a configuration. Plus, there is list in a dependency. We can probably just call:
```
from aind_video_utils.transcode import VIDEO_EXTENSIONS
```

File: src/aind_behavior_video_transformation/etl.py [hunk 11]
```diff
@@ -46,6 +48,68 @@ def _format_ffmpeg_error(video_path: Path, exc: CalledProcessError) -> str:
     )
 
 
+def encode_cost(path: Path) -> int:
```
Comment (jtyoung84) [inline src/aind_behavior_video_transformation/etl.py:51]: Maybe we can add the encode_cost in a future PR. For now, we can just partition the video files without balancing them.

File: src/aind_behavior_video_transformation/etl.py (hunk 10 above)
Comment (Hilda-Azimi) [reply src/aind_behavior_video_transformation/etl.py]: Done! I added "from aind_video_utils.transcode import VIDEO_EXTENSIONS" to both .py codes.

File: src/aind_behavior_video_transformation/etl.py (hunk 11 above)
Comment (Hilda-Azimi) [reply src/aind_behavior_video_transformation/etl.py]: Done!

**Comment by Hilda-Azimi** [comment]
Made the following changes based on the feedback:
1. Simplified video distribution: Videos and non-videos are now spread round-robin across partitions by sorted path, keeping the PR scoped to partitioning.
2. Removed the duplicated VIDEO_EXTENSIONS from filesystem.py and now import it from aind_video_utils as the single source of truth. 
3. Replaced the manual partition_number/num_partitions check in run_job with a Pydantic model_validator on the settings class, so an invalid config now fails at construction rather than at runtime.
4. Updated the partitioning tests to match, fixed the failures on Windows.

File: src/aind_behavior_video_transformation/etl.py [hunk 12]
```diff
@@ -201,7 +252,43 @@ def run_job(self) -> JobResponse:
             overrides,
```
Comment (jtyoung84) [inline src/aind_behavior_video_transformation/etl.py]: The `transform_directory` method being called on line 251 is the method that is creating the symlinks for the non-video files. This happens before the file list gets partitioned and is causing bugs. I ran some tests on the HPC wth 4 worker nodes against a folder with 4 video files and 4 csv files and this error shows up on one of the worker nodes:
```
Traceback (most recent call last):
  File "<frozen runpy>", line 203, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "/usr/local/lib/python3.13/site-packages/aind_behavior_video_transformation/etl.py", line 321, in <module>
    job_response = job.run_job()
  File "/usr/local/lib/python3.13/site-packages/aind_behavior_video_transformation/etl.py", line 248, in run_job
    convert_video_args = transform_directory(
        job_in_dir_path,
    ...<3 lines>...
        file_filter,
    )
  File "/usr/local/lib/python3.13/site-packages/aind_behavior_video_transformation/filesystem.py", line 137, in transform_directory
    symlink(file_path, out_path)
    ~~~~~~~^^^^^^^^^^^^^^^^^^^^^
FileExistsError: [Errno 17] File exists: 'ephys_rig_behavior_transfer/323_EPHYS3/858398/behavior_858398_2026-08-04_13-48-57/behavior-videos/BodyCamera/metadata.csv' -> 'test_output/BodyCamera/metadata.csv'
srun: error: n268: task 0: Exited with exit code 1
```
and these warnings show up on the other worker nodes:
```
WARNING:root:Output path test_output/BodyCamera/metadata.csv already exists!
WARNING:root:Output path test_output/BottomCamera/metadata.csv already exists!
WARNING:root:Output path test_output/SideCameraLeft/metadata.csv already exists!
WARNING:root:Output path test_output/SideCameraRight/metadata.csv already exists!
```
We'll need to move the line that is creating the symlinks until after the partitions are built.

**Review by jtyoung84 — APPROVED** [review]

**Comment by galenlynch** [comment]
Removing `parallel_compression` is a breaking change for existing job settings: `BasicJobSettings` is `extra="forbid"`, so a stored settings JSON that still carries the field now fails validation rather than ignoring it. The README's docker example on this branch still passes "parallel_compression": true, so that command no longer runs.

Maybe we should keep the field as a deprecated no-op with a warning. Either that or drop it and update the README example in the same PR. Either way the README needs an edit.

**Comment by Hilda-Azimi** [comment]
I updated the README. And you're right that this is a breaking change — if someone doesn't notice, their existing config will fail validation. So the version should go from 1.x to 2.0.0, which means semantic-release needs a BREAKING CHANGE: footer in the commit message to pick that up. I'll talk to Jon about how we want to handle it.
